discojs/src/models/gpt/layers.spec.ts

```diff
@@ -0,0 +1,308 @@
+  
+      // get predictions from both models
+      const output1 = model1.predict(input) as tf.Tensor;
+      const output2 = model2.predict(input) as tf.Tensor;
+  
+      const arr1 = await output1.data();
+      const arr2 = await output2.data();
+  
+      // check lengths are equal
+      expect(arr1.length).to.equal(arr2.length);
+  
+      // check that the models produce the same output
+      expect(arr1).to.deep.equal(arr2);
+
+      // Check that there are no NaN values in the outputs.
+      for (let i = 0; i < arr1.length; i++) {
+        expect(isNaN(arr1[i])).to.be.false;
+      }
+      for (let i = 0; i < arr2.length; i++) {
+        expect(isNaN(arr2[i])).to.be.false;
+      }
+
+    });
+  });
+
+  // CausalSelfAttention Layer tests
+  describe('CausalSelfAttention Helper Methods', function () {
+  
+    const config: CausalSelfAttentionConfig = {
+      name: 'testCSA',
+      contextLength: 5,
+      nHead: 2,
+      nEmbd: 8,          // divisible by nHead, so head size = 4
+      dropout: 0.0,      // no dropout for deterministic tests
+      nLayer: 2,
+      seed: 42
+    };
+  
+    let csa: CausalSelfAttention;
+  
+    // new instance of CausalSelfAttention before each test
+    beforeEach(() => {
+      csa = new CausalSelfAttention(config);
+      // dummy input has shape [batch, T, nEmbd] = [1, contextLength, nEmbd].
+      const dummyInput = tf.zeros([1, config.contextLength, config.nEmbd], 'float32');
+      csa.apply(dummyInput);
+    });
+  
+    afterEach(() => {
+      tf.disposeVariables();
+    });
+  
+    // describe('_dense', function () {
+    //   it('should compute x * kernel + bias correctly using addWeight', async function () {
+    //     const x = tf.tensor2d([[1, 2]], [1, 2]);
+    //     const kernel = csa.addWeight(
+    //       'dense_test_kernel',
+    //       [2, 2],
+    //       'float32',
+    //       tf.initializers.constant({ value: [[1, 0], [0, 1]] })
+    //     ) as tf.layers.LayerVariable;
+    //     const bias = csa.addWeight(
+    //       'dense_test_bias',
+    //       [2],
+    //       'float32',
+    //       tf.initializers.constant({ value: [0.5, -0.5] })
+    //     ) as tf.layers.LayerVariable;
+  
+    //     const output = csa._dense(x, kernel, bias);
+    //     const outData = await output.data();
+    //     // Expected calculation:
+    //     // [1,2] dot [[1,0],[0,1]] = [1,2] and then add bias [0.5, -0.5] gives [1.5, 1.5]
+    //     expect(Array.from(outData)).to.deep.equal([1.5, 1.5]);
+    //   });
+    // });
+  
+    describe('splitHeads', function () {
+      it('should reshape and transpose the input correctly', function () {
+        const B = 2;
+        const T = 6;
+        const totalChannels = config.nEmbd; // 8 channels
+        // input tensor with shape [B, T, totalChannels]
+        const input = tf.tensor3d(new Array(B * T * totalChannels).fill(1), [B, T, totalChannels]);
```

Comment:
[`tf.ones` creates a tensor filled with ones](https://js.tensorflow.org/api/4.22.0/#ones)

```suggestion
        const input = tf.ones([B, T, totalChannels]);
```